### PR TITLE
Closed shard should never open new engine

### LINKED ISSUE #47060: [CI] FlushIT.testSyncedFlushWithConcurrentIndexing failure
The test has failed multiple times:
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+matrix-java-periodic/ES_BUILD_JAVA=openjdk12,ES_RUNTIME_JAVA=zulu11,nodes=general-purpose/225/console
and on different branches
https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+7.4+matrix-java-periodic/ES_BUILD_JAVA=openjdk12,ES_RUNTIME_JAVA=corretto11,nodes=general-purpose/48/console

I could not reproduce locally

```
./gradlew ':server:integTest' --tests "org.elasticsearch.indices.flush.FlushIT.testSyncedFlushWithConcurrentIndexing" \
  -Dtests.seed=68566A4079D35F31 \
  -Dtests.security.manager=true \
  -Dtests.locale=rm-CH \
  -Dtests.timezone=Atlantic/Azores \
  -Dcompiler.java=12 \
  -Druntime.java=11

```

### PR DESCRIPTION
**Pull request by dnhatn**
We should not open new engines if a shard is closed. We break this assumption in #45263 where we stop verifying the shard state before creating an engine but only before swapping the engine reference. We can fail to snapshot the store metadata or checkIndex a closed shard if there's some IndexWriter holding the index lock.

Closes #47060

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 1]
```diff
@@ -3851,6 +3875,7 @@ public InternalEngine recoverFromTranslog(TranslogRecoveryRunner translogRecover
         shard.acquireAllReplicaOperationsPermits(shard.getOperationPrimaryTerm(), shard.getLastKnownGlobalCheckpoint(), 0L,
             ActionListener.wrap(r -> {
                 try (r) {
+                    resetWasStarted.set(true);
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: It looks like we rely on timing to hit the cases where reset has started and has not started. I think I would prefer explicit latches/barriers to more directly dictate the order?

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 2]
```diff
@@ -3914,6 +3949,7 @@ public InternalEngine recoverFromTranslog(TranslogRecoveryRunner translogRecover
...
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: It looks like we rely on timing to hit the cases where reset has started and has not started. I think I would prefer explicit latches/barriers to more directly dictate the order?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 3]
```diff
@@ -1309,20 +1307,22 @@ public CacheHelper getReaderCacheHelper() {
     }
 
     public void close(String reason, boolean flushEngine) throws IOException {
-        synchronized (mutex) {
-            try {
-                changeState(IndexShardState.CLOSED, reason);
-            } finally {
-                final Engine engine = this.currentEngineReference.getAndSet(null);
+        synchronized (engineMutex) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: I think this means that we now risk waiting for an engine warmer during close? I.e., partially undo the effect of #45263?

File: server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java [hunk 4]
```diff
@@ -4126,7 +4162,7 @@ protected void ensureMaxSeqNoEqualsToGlobalCheckpoint(SeqNoStats seqNoStats) {
         closeShards(readonlyShard);
     }
 
-    public void testCloseShardWhileEngineIsWarming() throws Exception {
+    public void testEngineWarmingDoesNotBlockClusterUpdate() throws Exception {
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/shard/IndexShardTests.java]: I think we want both cluster state updating and closing to be possible while warming?

**Review by henningandersen — COMMENTED** [review]
Thanks for looking into this @dnhatn, I believe I know what you want to achieve, but I also think this might revert back blocking the cluster state applier thread (more details in comments)?

**Comment by dnhatn** [comment]
@henningandersen Thank you for your thoughtful review.

> I think this means that we now risk waiting for an engine warmer during close.

Great catch. I spent some time on this, but I did not come up with a solution unless we introduce the "start" method to Engine. Do you have any suggestions for this?

**Comment by henningandersen** [comment]
@dnhatn I also lean towards the "start" method that runs outside the mutex. The other options I can think of are:

* Move the locking to IndexService to prevent snapshot'ing from hitting this (and fix checkIndex to also avoid it).
* Make close async, i.e., it reports back through a listener when done.
* Keep close, but fix the caller to run in another thread and async-notify.

Unfortunately, using the generic thread pool for the last approach might lead to issues if it can be called during shutdown?

I think I prefer to add a method to do the warmup, seems simpler overall. The benefit of doing close async though is that it avoids waiting for the rest of the IO happening during InternalEngine constructor. Not sure if this has any benefits?

**Comment by dnhatn** [comment]
> I think I prefer to add a method to do the warmup, seems simpler overall.

+1. Thank you for your suggestion. I will work on this solution.

**Comment by dnhatn** [comment]
@henningandersen I worked on a change that moves the engine warming out of the constructor. It is pretty straightforward. However, it does not eliminate the blocking issue. Closing an engine acquires the [writeLock](https://github.com/elastic/elasticsearch/blob/a8a7477c396554926f260d210364f009d85ae5f2/server/src/main/java/org/elasticsearch/index/engine/Engine.java#L1714), which can be blocked by an engine warming as it holds the readLock (via [refresh](https://github.com/elastic/elasticsearch/blob/a8a7477c396554926f260d210364f009d85ae5f2/server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java#L1734)). We can fix the refresh, but indexing and flushing can cause the same problem. I will reach out to discuss this with you.

**Comment by dnhatn** [comment]
I have two tests in https://github.com/elastic/elasticsearch/pull/47186/commits/a56d9ff2a6cfac17a728bc0bc4d6f32f2d6aaf90 which can reliably reproduce the test failure reported in #47060. However, neither of them works with the latest change as we now hold engineMutex while closing a shard. I am not sure if I can come up with a useful test for this change. Any suggestions would be great. 

@henningandersen It's ready. Can you please take another look? Thank you.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 5]
```diff
@@ -1393,7 +1393,7 @@ public long recoverLocallyUpToGlobalCheckpoint() {
                 getEngine().recoverFromTranslog(translogRecoveryRunner, globalCheckpoint);
                 logger.trace("shard locally recovered up to {}", getEngine().getSeqNoStats(globalCheckpoint));
             } finally {
-                synchronized (mutex) {
+                synchronized (engineMutex) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Maybe assert at the beginning of method that we do not hold `mutex`.

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 6]
```diff
@@ -1615,7 +1607,7 @@ private void onNewEngine(Engine newEngine) {
      * called if recovery has to be restarted after network error / delay **
      */
     public void performRecoveryRestart() throws IOException {
-        synchronized (mutex) {
+        synchronized (engineMutex) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: Maybe assert that we are not holding `mutex`?

**Review by henningandersen — APPROVED** [review]
LGTM. Thanks for the large effort on this, @dnhatn .

**Comment by dnhatn** [comment]
@henningandersen Thank you so much for your review and discussion on this.

**Pull request merged**